--- unbound-1.6.1/Makefile.in.orig	2016-12-06 16:29:35.000000000 +0100
+++ unbound-1.6.1/Makefile.in	2017-03-09 18:32:30.802012387 +0100
@@ -412,7 +412,7 @@
 	if test "$(LEX)" != ":"; then \
 		echo "#include \"config.h\"" > $@ ;\
 		echo "#include \"util/configyyrename.h\"" >> $@ ;\
-		$(LEX) -t $(srcdir)/util/configlexer.lex >> $@ ;\
+		$(LEX) -t --prefix=ub_c_ $(srcdir)/util/configlexer.lex >> $@ ;\
 	fi
 
 util/configparser.c util/configparser.h:  $(srcdir)/util/configparser.y
